optaplanner-examples/src/test/java/org/optaplanner/examples/conferencescheduling/optional/score/ConferenceSchedulingConstraintProviderTest.java

```diff
@@ -0,0 +1,919 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.conferencescheduling.optional.score;
+
+import static java.util.Collections.emptySet;
+import static java.util.Collections.singleton;
+import static java.util.Collections.singletonList;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneId;
+import java.util.Arrays;
+import java.util.HashSet;
+
+import org.junit.jupiter.api.Test;
+import org.optaplanner.examples.conferencescheduling.domain.ConferenceConstraintConfiguration;
+import org.optaplanner.examples.conferencescheduling.domain.ConferenceSolution;
+import org.optaplanner.examples.conferencescheduling.domain.Room;
+import org.optaplanner.examples.conferencescheduling.domain.Speaker;
+import org.optaplanner.examples.conferencescheduling.domain.Talk;
+import org.optaplanner.examples.conferencescheduling.domain.Timeslot;
+import org.optaplanner.test.api.score.stream.ConstraintVerifier;
+
+public class ConferenceSchedulingConstraintProviderTest {
+
+    private final ConstraintVerifier<ConferenceSchedulingConstraintProvider, ConferenceSolution> constraintVerifier =
+            ConstraintVerifier.build(new ConferenceSchedulingConstraintProvider(), ConferenceSolution.class,
+                    Talk.class);
+
+    private static final LocalDateTime START = LocalDateTime.ofInstant(Instant.EPOCH, ZoneId.systemDefault());
+
+    private static final Timeslot TIMESLOT1 = new Timeslot(1)
+            .withStartDateTime(START)
+            .withEndDateTime(START.plusHours(1))
+            .withTagSet(singleton("a"));
+    private static final Timeslot TIMESLOT2 = new Timeslot(2)
+            .withStartDateTime(TIMESLOT1.getEndDateTime())
+            .withEndDateTime(TIMESLOT1.getEndDateTime().plusHours(1))
+            .withTagSet(singleton("b"));
+    private static final Timeslot TIMESLOT3 = new Timeslot(3)
+            .withStartDateTime(TIMESLOT2.getEndDateTime())
+            .withEndDateTime(TIMESLOT2.getEndDateTime().plusHours(1))
+            .withTagSet(singleton("c"));
+    private static final Timeslot TIMESLOT_ANOTHER_DAY = new Timeslot(3)
+            .withStartDateTime(TIMESLOT1.getStartDateTime().plusDays(1))
+            .withEndDateTime(TIMESLOT1.getStartDateTime().plusDays(1).plusHours(1))
+            .withTagSet(singleton("c"));
```

Comment:
Potential missing test coverage:
The conf scheduling example allows for this:
- Timeslot A is from 9:00 to 10:00 with talk type "breakout".
- Timeslot B is from 9:20 to 11:20 with talk type "lab".
Notice how the first is 60 minutes, the second 120 minutes and they overlap for 40 minutes.

I don't see that kind of complexity in these timeslots as far as I call tell.

Some of the constraints will penalize based on the overlapping time, so by a factor of 40.
In that case it should be asserted that the matchPenality is 40, not 60 or 120 or 160 or 180.
Other constraints will penalize on the union time, so 160, or the summed time, so 180.